[Flang][OpenMP] Allow copyprivate and nowait on the directive clauses

Thirumalai-Shaktivel opened the pull request:
Issue:
- Single construct used to throw semantic error for copyprivate and nowiat clause when used in the single directive.
- Also, the Nowait and copyprivate restrictions has be removed from OpenMP 6.0

Fix:
- The above mentioned semantic error was valid until OpenMP 5.1.
- From OpenMP 5.2, copyprivate and nowait is allowed in both clause and end-clause

From Reference guide (OpenMP 5.2, 2.10.2):
```
!$omp single [clause[ [,]clause] ... ]
loosely-structured-block
!$omp end single [end-clause[ [,]end-clause] ...]

clause:
  copyprivate (list)
  nowait
  [...]

end-clause:
  copyprivate (list)
  nowait
```

Towards: https://github.com/llvm/llvm-project/issues/110008

Thirumalai-Shaktivel committed:
[Flang][OpenMP] Allow copyprivate and nowait on the directive clauses

Issue:
- Single construct used to throw semantic error for copyprivate and
  nowiat clause when used in the single directive.
- Also, the Nowait and copyprivate restrictions has be removed from
  OpenMP 6.0

Fix:
- The above mentioned semantic error was valid until OpenMP 5.1.
- From OpenMP 5.2, copyprivate and nowait is allowed in both
  clause and end-clause

From Reference guide (OpenMP 5.2, 2.10.2):
```
!$omp single [clause[ [,]clause] ... ]
loosely-structured-block
!$omp end single [end-clause[ [,]end-clause] ...]

clause:
  copyprivate (list)
  nowait
  [...]

end-clause:
  copyprivate (list)
  nowait
```

kiranchandramohan commented:
Would it make sense to allow this usage irrespective of the standard?

kkwli commented:
> Would it make sense to allow this usage irrespective of the standard?

My two cents. If flang supports both forms, it'd better to disallow mixing two forms in a **single** construct. There is if **copyprivate** is specified on the **single** directive, there shouldn't be anything on the **end single**. I guess that is less confusing.

Thirumalai-Shaktivel commented:
> Would it make sense to allow this usage irrespective of the standard?

Yes, that makes sense, I will make the required changes soon.


@kkwli, I have added a check to catch the duplicate usage of the clauses, see: https://github.com/llvm/llvm-project/pull/127769/files#diff-69cb55326952c1dffb012c3c2691d8ddc1e65bf9500fd41821f1196d04216bf9R39-R48. I hope, it looks good.

Thirumalai-Shaktivel committed:
[Flang] Remove semantic checks for Copyprivate and nowait

Thirumalai-Shaktivel committed:
Fix failures

Thirumalai-Shaktivel committed:
Remove duplicate checks

Thirumalai-Shaktivel commented:
Ready for review!

Thirumalai-Shaktivel commented:
Ready for review!

kkwli commented:
> > > Would it make sense to allow this usage irrespective of the standard?
> > 
> > 
> > My two cents. If flang supports both forms, it'd better to disallow mixing two forms in a **single** construct. There is if **copyprivate** is specified on the **single** directive, there shouldn't be anything on the **end single**. I guess that is less confusing.
> 
> I would argue that this should be warning instead of a hard error. The code would not be incorrect according to the OpenMP API and therefore should be accepted, and the warning should tell the user to be consistent.

I fine with lowering to a warning.

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -1203,6 +1203,39 @@ void OmpStructureChecker::Enter(const parser::OpenMPBlockConstruct &x) {
+    auto catchCopyPrivateNowaitClauses = [&](const auto &dir) {
+      for (auto &clause : std::get<parser::OmpClauseList>(dir.t).v) {
+        if (clause.Id() == llvm::omp::Clause::OMPC_copyprivate) {
+          if (foundCopyPrivate) {
```
I missed this, thanks for noticing.

Thirumalai-Shaktivel replied:
Done

Thirumalai-Shaktivel committed:
Fix Copyprivate clause semantics check

Changes:
- Throw error for repeated list items in the copyprivate clause
- Throw warnings for repeated use of copyprivate and nowait in the
  end directive

Thirumalai-Shaktivel committed:
Add a test

Thirumalai-Shaktivel commented:
@mjklemm, I made the required changes, I have added some more testcases, do you see anything missing there.
Also, I have a doubt, consider an example:
```fortran
program single
    !$omp single nowait
        print *, x
    !$omp end single copyprivate(x, y) nowait
end program
```
Error:
```console
$ flang x.f90 -fopenmp
flang-21: warning: OpenMP support in flang is still experimental [-Wexperimental-option]
error: Semantic errors in x.f90
./x.f90:2:18: error: NOWAIT clause must not be used with COPYPRIVATE clause on the SINGLE directive
      !$omp single nowait
                   ^^^^^^
./x.f90:4:40: warning: NOWAIT clause is already used on the SINGLE directive
      !$omp end single copyprivate(x, y) nowait
                                         ^^^^^^
```
Is the error and the warning correct?

Thirumalai-Shaktivel committed:
Fix clang format

mjklemm commented:
> Error:
> 
> ```
> $ flang x.f90 -fopenmp
> flang-21: warning: OpenMP support in flang is still experimental [-Wexperimental-option]
> error: Semantic errors in x.f90
> ./x.f90:2:18: error: NOWAIT clause must not be used with COPYPRIVATE clause on the SINGLE directive
>       !$omp single nowait
>                    ^^^^^^
> ./x.f90:4:40: warning: NOWAIT clause is already used on the SINGLE directive
>       !$omp end single copyprivate(x, y) nowait
>                                          ^^^^^^
> ```
> 
> Is the error and the warning correct and expected?

I think this is a bug.  I do not see any restriction in the OpenMP specification that would disallow NOWAIT when COPYPRIVATE is present.

However, the compiler should produce an error that you have NOWAIT at both the begin and end directive.  Only for the COPYPRIVATE clause the diagnostic should be a warning.  I consider NOWAIT at both directives an actual error, because NOWAIT is a unique clause as per [5.2:308:10].

Thirumalai-Shaktivel commented:
Thanks for the review!

> I think this is a bug. I do not see any restriction in the OpenMP specification that would disallow NOWAIT when COPYPRIVATE is present.

Maybe I think I need to fix the error message there.
Here is the restriction from OpenMP 5.2: 11.1 single Construct, which says "The copyprivate clause must not be used with the nowait clause". But this restriction is removed in OpenMP 6.0.

I actually altered the previous error message, it used to throw the following:
```console
error: Semantic errors in <source>
<source>:7:7: error: Clause NOWAIT is not allowed if clause COPYPRIVATE appears on the END SINGLE directive
  !$omp end single copyprivate(x) nowait
        ^^^^^^^^^^
```
Let me know, what needs to be done here.

> However, the compiler should produce an error that you have NOWAIT at both the begin and end directive. Only for the COPYPRIVATE clause the diagnostic should be a warning. I consider NOWAIT at both directives an actual error, because NOWAIT is a unique clause as per [5.2:308:10].

This seems right, I will make the required changes soon.

Thirumalai-Shaktivel committed:
Fix: Allow nowait only once

Thirumalai-Shaktivel commented:
Done, now I get: 
```console
$ flang-new x.f90 -fopenmp              
flang-21: warning: OpenMP support in flang is still experimental [-Wexperimental-option]
error: Semantic errors in x.f90
./x.f90:2:18: error: NOWAIT clause must not be used with COPYPRIVATE clause on the SINGLE directive
      !$omp single nowait
                   ^^^^^^
./x.f90:4:40: error: At most one NOWAIT clause can appear on the SINGLE directive
      !$omp end single copyprivate(x, y) nowait
                                         ^^^^^^
```

mjklemm reviewed (approved):
I think this LGTM.

Thirumalai-Shaktivel commented:
Thanks for the review!
